### ISSUE TITLE
Add Helper functions for anonymizing Order (Payment) data

### ISSUE THREAD

**Issue by cklosowski** [easydigitaldownloads/easy-digital-downloads#6497]
## Enhancement Request
EDD needs some helper functions centered around order (payment) data to facilitate anonymizing the unnecessary data, while keeping in mind that stores may need to retain certain data to remain compliant with local tax and financial laws.

### Justification or use case
This is a [GDPR](#6216 ) issue. While core hasn't finished their tools yet, we should start building our helper functions and making sure things don't break when used, so that when core is ready we have working functions/methods.

**Comment by cklosowski** [comment]
I left this on the Pull Request, but for people looking over issues, here is the lowdown on what this does:

- Adds support for payment records to be anonymized or deleted.
- Adds `_edd_anonymize_payment` to handle anonymizing or deleting a payment record

There is a filter in place associated with the status of a payment. By default these are the actions taken:
```
case 'publish':
case 'refunded':
case 'revoked':
	$action = 'anonymize';
	break;

case 'failed':
case 'abandoned':
	$action = 'delete';
	break;

case 'pending':
case 'processing':
default:
	$action = 'none';
	break;
```

TL;DR - Published, refunded, or revoked get anonymized, failed and abandoned get deleted, and pending or processing have no action taken as they are still possibly needed. No action is also the default.

The filter `edd_privacy_payment_status_action_' . $action` exists so that extensions and developers with custom statuses can choose what to do with their status (thinking recurring and Stripe off hand).

The data that is anonymized when that action is taken is:

- First Name is made blank
- Last  Name is made blank
- All email addresses are converted to the anonymized email address on the customer
- The IP address is run to only be the /24 IP Address (ending in .0) so it cannot be traced back to a user
- Address line 1 is made blank
- Address line 2 is made blank

**Comment by cklosowski** [comment]
I'm going to add another step in here, that runs after the payment delete, anonymizer. This one will check that after all payments are handled, that if there are no customers associated with this email address anymore, to delete the customer.

**Comment by Spreeuw** [comment]
Thank you so much for the work so far! I think the deletion/anonymizing should be a setting somewhere, because the payment records also serve as accounting record in many stores, and deleting this would break compliancy with other EU laws that overrule GDPR.

**Comment by cklosowski** [comment]
@spreeuw Thanks for the feedback. I'll look into a setting where store owners can choose what to do with certain statuses. We do have filters in place for things like Simple Shipping so a completed order that still has to be shipped, isn't anonymized.

Keep in mind that I'm not anonymizing nearly anything that would negate tax laws. All that's being removed are names and direct addresses. With the IP in a /24, the zip, city, state, country taxes can usually be determined still.

I'll look into that though, once I get this through a couple tests to make sure its working as expected first.

@easydigitaldownloads/core-team I'm going to add a few more things in there.

1) After this runs I'm going to see if the customer requesting deletion has any payments (since they could have been all deleted if they were all abandoned or failed) and delete the customer at that point.
2) See if I can work up some settings for store owners to choose what to do with specific statuses, keeping in mind that some extensions might override that, like the case of completed but not shipped (with simple shipping).

**Comment by cklosowski** [comment]
So it looks like the first isn't possible, because we have to anonymize the customer prior to anonymizing the payment...so now we can no longer look up the customer anymore, b/c it has been anonymized.

I'll go through with the 2nd item though.

**Comment by cklosowski** [comment]
Settings added, and with that, I've also added in that if a customer has payments that are not eligible for anonymization, the customer is not anonymzied. This would be things like payments that are processing, pending, etc.

**Comment by SDavisMedia** [comment]
Initial tests are good here. The payment meta is being removed/anonymized as expected along with the associated customers. So this is working as expected. :+1:

However, I want to test the new Privacy Settings options for the different payment statuses to make sure they do what is expected in default state as well as when they have different selections. So I'll get to those tests shortly.

**Comment by SDavisMedia** [comment]
Already discussed in Slack, but my initial test of the privacy settings indicate that the settings aren't being respect. With the following settings:

![image 2018-05-18 at 9 01 30 pm](https://user-images.githubusercontent.com/2359131/40274003-f92f9102-5b80-11e8-84fd-3b2bdcf44cde.png)

I got the following results while attempting to erase a customer with at least one payment record for every type of payment status:

![image 2018-05-18 at 9 05 37 pm](https://user-images.githubusercontent.com/2359131/40274010-0d83b55c-5b81-11e8-9ed3-91c3f440c5e4.png)

It makes sense that the customer would not be anonymized, considering there are payments that affect whether that should happen or not. But the payments themselves were affected based on the default behavior of the eraser in regard to the payment statuses, as opposed to what I chose in my privacy settings.

Chris is aware and is working on addressing that. I'll test again once that's done.
